kep/219-pod-allocatable-ip-scheduling/README.md

```diff
@@ -0,0 +1,49 @@
+- [Design Details](#design-details)
+- [Implementation History](#implementation-history)
+<!-- /toc -->
+
+## Summary
+
+This plugin takes node's `.spec.PodCIDR` into account in scheduling cycle. When the density of pods in a node is relatively high, nodes might not have sufficient allocatable Pod IP to allocate.
+Even though kubelet has its own `--max-pods` option flag to limit the total number of Pods on nodes, we can filter those nodes who have insufficient Pod IP to allocate in scheduling cycle. 
+
+## Motivation
+[kube-controller-manager](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager/) provides the `--allocate-node-cidrs` options to mark Pod CIDR in each node's `.spec.PodCIDR` field.
+When a node has already allocated all the Pod IPs from `.spec.PodCIDR`, a scheduling Pod will ignore this node in scheduling cycle. It avoids the situation that pods are binding to insufficient Pod IP nodes,
+which is useful when a Pod cannot be scheduled due to the limited Pod IP allocation constraint in one node. This constraint is likely to result in Pods stucking in `ContainerCreating` phase.
+
+### Goals
+
+1. Implement a Filter plugin to filter out the nodes who have already allocated all the Pod IPs from `.spec.PodCIDR`.
+2. This plugin only works for those clusters who use standard CNI plugins and enable the `--allocate-node-cidrs` options in [kube-controller-manager](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager/), which means all pod IPs are strictly allocated from each node's `.spec.PodCIDR`.
+
+### Non-Goals
+
+1. Considering each node's Pod Allocatable IP in those clusters who use Elastic Network Interface(ENI) for Pod.
+2. Considering each node's Pod Allocatable IP in those clusters whose `--allocate-node-cidrs` options of kube-controller-manager is disabled, which means the PodCIDR is not recorded into node resources.
+
+## Use Cases
+
+Make sure to set the `--allocate-node-cidrs=true --cluster-cidr=<cidr>` options in kube-controller-manager and use standard [CNI plugins](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/network-plugins/). Each node would be assigned an IP subnet through the `--cluster-cidr=<cidr>` configuration.
+For the node resources, `.spec.PodCIDR` field will be set as well.
+
+## Design Details
+
+We implement a filter plugin based on the scheduler framework. 
+
+After nodes were joining into the cluster, their Pod IP subnet was assigned, and the `.spec.PodCIDR` field in node resources was set as well.
+
+When a Pod is in scheduling cycle, the Pod Allocatable IP Filter Plugin will check if nodes has sufficient allocatable Pod IP to assign according its `.spec.PodCIDR` field.
+
+Comparing each node's `.spec.PodCIDR` field and Pod IPs that have already been assigned, the Plugin calculates the number of allocatable Pod IP, filters out those nodes who have insufficient allocatable Pod IP to assign.
```

Comment:
How to calculate the number of allocatable IPs? Please add details.

Reply:
Questions. 
1. When and how is a node's `spec.PodCIDR` updated? 
2. Will race conditions be possible?  
 - A node is considered feasible in filtering/scoring phases, but its allocatable IPs is updated in binding and late phases. 
 - A node is chosen for a pod in a scheduling cycle.  Before the pods is bound or created, `spec.PodCIDR` will not be updated and its value won't reflect the accurate allocatable IPs on the node. The next scheduling cycle will mis-calculate the allocatable IPs on the node . 

Do you need additional plugins (Permit/Reserve/Unreserve/PostBind) to handle IP state?

Reply:
> When and how is a node's spec.PodCIDR updated?

I think the node's `spec.PodCIDR` updated only when users manually modify the field to change the ip pool for pod, or `--cluster-cidr` has been modified in kube-controller-manager. Under both circumstance, pods needed to be re-created and pod IPs needed to be re-assigned as well.

Reply:
> 2. Will race conditions be possible?
> 
> * A node is considered feasible in filtering/scoring phases, but its allocatable IPs is updated in binding and late phases.
> * A node is chosen for a pod in a scheduling cycle.  Before the pods is bound or created, `spec.PodCIDR` will not be updated and its value won't reflect the accurate allocatable IPs on the node. The next scheduling cycle will mis-calculate the allocatable IPs on the node .
> 
> Do you need additional plugins (Permit/Reserve/Unreserve/PostBind) to handle IP state?

Yes, this is a good suggestion, I think we need an extra PreBind plugin to double check the allocatable IPs. This will make sure the number of allocatable IPs  is accurate before the binding.
@yuanchen8911

Reply:
> How to calculate the number of allocatable IPs? Please add details.

done. Thanks for the suggestions and reviews!
@yuanchen8911

Reply:
> > When and how is a node's spec.PodCIDR updated?
> 
> @yuanchen8911 I think the node's `spec.PodCIDR` updated only when users manually modify the field to change the ip pool for pod, or `--cluster-cidr` has been modified in kube-controller-manager. Under both circumstance, pods needed to be re-created and pod IPs needed to be re-assigned as well.

1. When is a new pod's `spec.PodCIDR` updated,  binding or creating phase? Which controller or scheduler does it? 
2. When an existing pod is deleted, which controller is responsible for updating the field, e.g., reclaiming the IP and making it allocatable?

Reply:
> When is a new pod's spec.PodCIDR updated, binding or creating phase? Which controller or scheduler does it?

`spec.PodCIDR` is a static field in node resources to record Pod IP subnet, it won't be updated after assigning IPs for pod. 

> When an existing pod is deleted, which controller is responsible for updating the field, e.g., reclaiming the IP and making it allocatable?

As what I was mentioned in ` Design Details ` section, the plugin starts a client to list all the Pods that has been scheduled in this node, which can calculate the accurate allocatable IPs. ( There is no need to concern about the  ip reclaiming or assigning mechanism)

@yuanchen8911

Reply:
OK，I guess I misunderstood the meaning of the field. It specifies the maximum number of allocable IPs of a node, which is set statically. Right?

Reply:
Yes, it is set statically.
@yuanchen8911

Reply:
I'm afraid a node's allocated IPs is not equivalent to the number of running pods from scheduling perspective. The scheduler should also consider pods that are successfully scheduled but not created yet. Otherwise, potential race condition can happen as mentioned in a previous comment. 

Once the scheduler successfully schedules a pod and moves it to binding phase, allocated IPs should be deducted by 1 because the a new scheduling cycle will start. 

Once a pod is created and starts running in a node, the scheduler needs to update the allocated IPs. Otherwise, allocated IPs will be double counted. 

Another case is if a scheduled pod is stuck in containerCreating, When should its allocated IP be recycled? What about if it's deleted before being created? 

So the scheduler needs to watch for pod addition and deletion events and update allocated IP count accordingly.

Reply:
> I'm afraid a node's allocated IPs is not equivalent to the number of running pods from scheduling perspective. The scheduler should also consider pods that are successfully scheduled but not created (running) yet. Otherwise, a race condition can happen as mentioned in a previous comment.
> 

The plugin is not just the considering the running pod. Actually, all the pods that has been bound (the pods' `nodeName` field were set ) should be taken into account.

> After the scheduler successfully schedules a pod and moves it to the binding phase, allocatable IPs should be deducted by 1 because the a new scheduling cycle will start and the scheduled pod is not running yet. Once the pod starts running, the scheduler needs to update the node's allocatable IPs too (plus 1). Otherwise, the allocated IPs will be double counted (i.e., IP leak).
> 

That's why we need a preBind plugin to check the node's allocatable IPs right before the pods' binding phase in the binding cycle.

> Another scenario is if a scheduled pod is stuck in containerCreating, when will its allocated IP be recycled? What about if the pod is deleted before it's created?

In this scenario, we consider this Pod has already taken a pod IP assignment quota from node (no need to recycle the allocated IP). Even though Pod IP has not been assigned to this pod yet, we assume the IP assignment will happen eventually as soon as other Pods are deleted. 
If a node has pods stuck in containerCreating for the insufficient ip reasons, no other pods can be scheduled to this node either.

> I think the scheduler will need to watch for pod addition and deletion events and update allocated or allocatable IP count accordingly.

If counting the number of those pods who have the same `nodeName` field value can prove how many pods have already bound to the same nodes, I don't think this is necessary.

@yuanchen8911

Reply:
Assume a node's IP capacity is 8 with 4 pods running and a new pod is submitted (will be placed on the node). Here are the allocated IP changes.

1. Before pod scheduling: 4
2. After pod scheduling: 5 (!=#running pods; update must be done before prebind because a new schedule cycle will start after that)
3. If the scheduled pod is deleted before running: 4;
If the pod is created and running: 5

Without tracking pod creation and deletion events, how can you tell which case it is and hence update allocated IPs differently for the next pod?

Reply:
> After pod scheduling and before pod creation: 5 (!= num. of running pods; 

If this pod has finished its binding cycle, the allocatable IP will be 3 (8-5=3).
If this pod has not finished binding to this node, the allocatable IP will still be 4 (8-4=4).
It doesn't matter if this pod is running or not, as long as this pod has finished bound to a node after a binding cycle, num of node's allocatable IPs will minus 1.

> update must be done before prebind (e.g. reservation) because a new schedule cycle will start after that)
> Without tracking pods' creation and deletion events, how can you tell which case it is and hence update allocated IPs differently?

we don't need a place to store each node's allocated IP, the preBind plugin checks the number of pods that have already been bound to the node ( only when a pod finished its binding cycle that can have its`.nodeName` field be set) to determine if this node has allocatable IPs to assign.

@yuanchen8911

Reply:
Maybe we need a discussion offline. 

Additional detail with examples about how to track IP count in the KEP would be helpful anyway.